airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java

```diff
@@ -153,24 +155,43 @@ private Thread getWorkerThread(Worker<INPUT, OUTPUT> worker, CompletableFuture<O
     });
   }
 
+
+  /**
+   * Cancel is implementation in a slightly convoluted manner due to Temporal's semantics. Cancel requests are routed to the Temporal Scheduler via
+   * the cancelJob function in SchedulerHandler.java. This manifests as a {@link io.temporal.client.ActivityCompletionException} when the {@link CancellationHandler}
+   * heartbeats to the Temporal Scheduler.
+   *
+   * The callback defined in this function is executed after the above exception is caught, and defines the clean up operations executed as part of cancel.
+   *
+   * See {@link CancellationHandler} for more info.
+   */
   private Runnable getCancellationChecker(Worker<INPUT, OUTPUT> worker, Thread workerThread, CompletableFuture<OUTPUT> outputFuture) {
+    var cancelled = new AtomicBoolean(false);
     return () -> {
       try {
         mdcSetter.accept(jobRoot, jobId);
 
         final Runnable onCancellationCallback = () -> {
+          if (cancelled.get()) {
+            // Since this is a separate thread, race condition between the executor service shutting down and this thread's next invocation can happen. This
+            // check guarantees cancel operations are only executed once.
+            return;
+          }
+
           LOGGER.info("Running sync worker cancellation...");
+          cancelled.set(true);
           worker.cancel();
 
           LOGGER.info("Interrupting worker thread...");
           workerThread.interrupt();
 
           LOGGER.info("Cancelling completable future...");
+          // This throws a CancellationException as part of the cancelling and is the exception seen when cancelling the job.
           outputFuture.cancel(false);
         };
 
         cancellationHandler.checkAndHandleCancellation(onCancellationCallback);
-      } catch (WorkerException e) {
+      } catch (Exception e) {
```

Comment:
Because the cancellation handler no longer throws a worker exception, this becomes a general exception in case we run into anything. we could get rid of this.